src/common/libjob/spec.h

```diff
@@ -0,0 +1,139 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#ifndef _FLUX_CORE_SPEC_H
+#define _FLUX_CORE_SPEC_H
+
+typedef struct flux_jobspec1 flux_jobspec1_t;
+
+/* Remove the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator. Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_attr_del (flux_jobspec1_t *jobspec, const char *path);
+
+/* Set the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * 'fmt' should be a jansson pack-style string corresponding to the
+ * remaining arguments.
+ * Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_attr_pack (flux_jobspec1_t *jobspec,
+                             const char *path,
+                             const char *fmt,
+                             ...);
+
+/* Unpack the values in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * 'fmt' should be a jansson unpack-style string corresponding to the
+ * remaining args.
+ * Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_attr_unpack (flux_jobspec1_t *jobspec,
+                               const char *path,
+                               const char *fmt,
+                               ...);
+
+/* Check the validity of the 'attributes' section of a jobspec, writing an error
+ * message to 'errbuf'.
+ * Return 0 on success, -1 on error with an error message written.
+ * This function succeeding does not guarantee that the jobspec is valid.
+ */
+int flux_jobspec1_attr_check (flux_jobspec1_t *jobspec, char *errbuf, int errbufsz);
```

Comment:
General question: is it possible to eliminate this and just have attr_pack and attr_del fail if the requested change would invalidate the attributes section?

Reply:
Definitely possible. The simplest thing to do would be to make a copy of the attributes section, apply the changes, run the existing ``attr_check`` function and then either return an error or apply the changes to the original attributes section. It wouldn't be very efficient but it wouldn't take very much code either.

I'm ambivalent about whether that would be an interface improvement, though. I'll defer to you.

Reply:
Feeling a bit ambivalent too.  Maybe what we should do is go ahead with the API as is, and expect to do a little further "co-design" with Cray once they begin to use it?  I have a feeling things will be much clearer once we have a real user.

Reply:
That seems like a good approach to me. I think we have a user who is ready and willing to start using it more or less immediately.